<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>API Reference · SeededAlignment.jl</title><meta name="title" content="API Reference · SeededAlignment.jl"/><meta property="og:title" content="API Reference · SeededAlignment.jl"/><meta property="twitter:title" content="API Reference · SeededAlignment.jl"/><meta name="description" content="Documentation for SeededAlignment.jl."/><meta property="og:description" content="Documentation for SeededAlignment.jl."/><meta property="twitter:description" content="Documentation for SeededAlignment.jl."/><meta property="og:url" content="https://MurrellGroup.github.io/SeededAlignment.jl/api/"/><meta property="twitter:url" content="https://MurrellGroup.github.io/SeededAlignment.jl/api/"/><link rel="canonical" href="https://MurrellGroup.github.io/SeededAlignment.jl/api/"/><script data-outdated-warner src="../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../search_index.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../">SeededAlignment.jl</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../">Home</a></li><li class="is-active"><a class="tocitem" href>API Reference</a><ul class="internal"><li><a class="tocitem" href="#Core-Methods"><span>Core Methods</span></a></li><li><a class="tocitem" href="#Utilities"><span>Utilities</span></a></li><li><a class="tocitem" href="#Core-Types"><span>Core Types</span></a></li><li><a class="tocitem" href="#Standard-Parameters"><span>Standard Parameters</span></a></li><li><a class="tocitem" href="#Index"><span>Index</span></a></li><li><a class="tocitem" href="#Index-Docstrings"><span>Index Docstrings</span></a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>API Reference</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>API Reference</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/MurrellGroup/SeededAlignment.jl" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/MurrellGroup/SeededAlignment.jl/blob/main/docs/src/api.md" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="API-Reference"><a class="docs-heading-anchor" href="#API-Reference">API Reference</a><a id="API-Reference-1"></a><a class="docs-heading-anchor-permalink" href="#API-Reference" title="Permalink"></a></h1><p>This section provides detailed documentation for all exported types and functions in <code>SeededAlignment.jl</code>.</p><h2 id="Core-Methods"><a class="docs-heading-anchor" href="#Core-Methods">Core Methods</a><a id="Core-Methods-1"></a><a class="docs-heading-anchor-permalink" href="#Core-Methods" title="Permalink"></a></h2><ul><li><strong><code>seed_chain_align</code></strong> </li><li><strong><code>msa_codon_align</code></strong> </li><li><a href="#SeededAlignment.clean_frameshifts-Tuple{BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}, BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}}"><code>clean_frameshifts</code></a></li><li><a href="#SeededAlignment.nw_align-Tuple{BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}, BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}}"><code>nw_align</code></a></li></ul><h2 id="Utilities"><a class="docs-heading-anchor" href="#Utilities">Utilities</a><a id="Utilities-1"></a><a class="docs-heading-anchor-permalink" href="#Utilities" title="Permalink"></a></h2><p>methods for reading and writing of sequence data in FASTA format.</p><ul><li><a href="#SeededAlignment.read_fasta-Tuple{String}"><code>read_fasta</code></a></li><li><a href="#SeededAlignment.write_fasta-Tuple{String, Tuple{BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}, BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}}}"><code>write_fasta</code></a></li></ul><h2 id="Core-Types"><a class="docs-heading-anchor" href="#Core-Types">Core Types</a><a id="Core-Types-1"></a><a class="docs-heading-anchor-permalink" href="#Core-Types" title="Permalink"></a></h2><ul><li><code>Move</code></li><li><code>Moveset</code></li><li><code>ScoringScheme</code></li><li><code>LongDNA{4}</code></li></ul><h2 id="Standard-Parameters"><a class="docs-heading-anchor" href="#Standard-Parameters">Standard Parameters</a><a id="Standard-Parameters-1"></a><a class="docs-heading-anchor-permalink" href="#Standard-Parameters" title="Permalink"></a></h2><ul><li><code>std_scoring</code></li><li><code>std_codon_moveset</code></li><li><code>std_noisy_moveset</code></li></ul><h2 id="Index"><a class="docs-heading-anchor" href="#Index">Index</a><a id="Index-1"></a><a class="docs-heading-anchor-permalink" href="#Index" title="Permalink"></a></h2><p>A complete alphabetical listing of all documented functions.</p><ul><li><a href="#SeededAlignment.Move"><code>SeededAlignment.Move</code></a></li><li><a href="#SeededAlignment.ScoringScheme"><code>SeededAlignment.ScoringScheme</code></a></li><li><a href="#SeededAlignment.clean_frameshifts-Tuple{BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}, BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}}"><code>SeededAlignment.clean_frameshifts</code></a></li><li><a href="#SeededAlignment.msa_codon_align-Tuple{BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}, Vector{BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}}}"><code>SeededAlignment.msa_codon_align</code></a></li><li><a href="#SeededAlignment.nw_align-Tuple{}"><code>SeededAlignment.nw_align</code></a></li><li><a href="#SeededAlignment.nw_align-Tuple{BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}, BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}}"><code>SeededAlignment.nw_align</code></a></li><li><a href="#SeededAlignment.read_fasta-Tuple{String}"><code>SeededAlignment.read_fasta</code></a></li><li><a href="#SeededAlignment.seed_chain_align-Tuple{BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}, BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}}"><code>SeededAlignment.seed_chain_align</code></a></li><li><a href="#SeededAlignment.std_codon_moveset-Tuple{}"><code>SeededAlignment.std_codon_moveset</code></a></li><li><a href="#SeededAlignment.std_noisy_moveset-Tuple{}"><code>SeededAlignment.std_noisy_moveset</code></a></li><li><a href="#SeededAlignment.std_scoring-Tuple{}"><code>SeededAlignment.std_scoring</code></a></li><li><a href="#SeededAlignment.write_fasta-Tuple{String, Tuple{BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}, BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}}}"><code>SeededAlignment.write_fasta</code></a></li><li><a href="#SeededAlignment.write_fasta-Tuple{String, Vector{BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}}}"><code>SeededAlignment.write_fasta</code></a></li></ul><h2 id="Index-Docstrings"><a class="docs-heading-anchor" href="#Index-Docstrings">Index Docstrings</a><a id="Index-Docstrings-1"></a><a class="docs-heading-anchor-permalink" href="#Index-Docstrings" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SeededAlignment.Move" href="#SeededAlignment.Move"><code>SeededAlignment.Move</code></a> — <span class="docstring-category">Type</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">Move</code></pre><p>Represents a single allowed alignment step between two sequences in the alignment matrix.</p><p><strong>Fields</strong></p><ul><li><code>step::Int</code>: How many aligned units (nucleotides) this move advances by.</li><li><code>score::Float64</code>: Cost or score associated with performing this move.</li></ul><p><strong>Reference to the top sequence (horizontal axis)</strong></p><ul><li><code>horizontal_stride::Int</code>: How far this move advances along the top (reference) sequence.</li><li><code>horizontal_phase::Int</code>: Frame offset in the top sequence; helps track codon boundaries.</li></ul><p><strong>Reference to the bottom sequence (vertical axis)</strong></p><ul><li><p><code>vertical_stride::Int</code>: How far this move advances along the bottom (query) sequence.</p></li><li><p><code>vertical_phase::Int</code>: Frame offset in the bottom sequence.</p></li><li><p><code>extensionAble::Bool</code>: Whether this move can be extended (e.g., for affine gap penalties).</p></li></ul><p><strong>Description</strong></p><p>Moves define basic operations used in dynamic programming alignment: matches, mismatches, and gaps. Moves can preserve codon reading frames by keeping strides in multiples of 3 and matching phase positions.</p><p>Two helper constructors are provided:</p><ul><li><code>Move(step, score, stride, phase, extensionAble)</code> assumes reference (vertical) stride and phase</li><li><code>Move(step, score, extensionAble)</code> for simple moves with unit stride and zero phase.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/MurrellGroup/SeededAlignment.jl/blob/38a40c0887ec27f271ae16cb2f51d79048aec487/src/alignmentOptions/moveset.jl#L2-L28">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SeededAlignment.ScoringScheme" href="#SeededAlignment.ScoringScheme"><code>SeededAlignment.ScoringScheme</code></a> — <span class="docstring-category">Type</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">ScoringScheme(; match_score=0.0, mismatch_score=0.5,extension_score=0.1,edge_ext_begin=true,edge_ext_end=true,kmerlength=21)</code></pre><p>A struct for storing scoring parameters used for sequence alignment. Uses a keyword constructor. </p><p><strong>Fields</strong></p><ul><li><p><code>match_score::Float64</code>: Score (typically ≤ 0) awarded for matching nucleotide. Lower is better.</p></li><li><p><code>mismatch_score::Float64</code>: Penalty for nucleotide mismatches. Higher values penalize substitutions more strongly.</p></li><li><p><code>extension_score::Float64</code>: Cost to extend a gap (indel). Affects how gaps are penalized during alignment.</p></li><li><p><code>edge_ext_begin::Bool</code>: If <code>true</code>, allows gaps to be extended at the <strong>beginning</strong> of sequences.</p></li><li><p><code>edge_ext_end::Bool</code>: If <code>true</code>, allows gaps to be extended at the <strong>end</strong> of sequences.</p></li><li><p><code>kmerlength::Int64</code>: Length of k-mers used for seeding alignments (if applicable). Ignored if no seeding is used.</p></li></ul><p><strong>Description</strong></p><p><code>ScoringScheme</code> defines how matches, mismatches, and gaps are scored during nucleotide-level sequence alignment. </p><p>This struct is typically passed to functions like <code>seed_chain_align</code> or <code>msa_codon_align</code>.</p><p><strong>example</strong></p><pre><code class="language-julia hljs">score_params = ScoringScheme(extension_score = 0.3, mismatch_score = 0.7) # (everything else will be keept at default values)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/MurrellGroup/SeededAlignment.jl/blob/38a40c0887ec27f271ae16cb2f51d79048aec487/src/alignmentOptions/scoring_scheme.jl#L1-L27">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SeededAlignment.clean_frameshifts-Tuple{BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}, BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}}" href="#SeededAlignment.clean_frameshifts-Tuple{BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}, BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}}"><code>SeededAlignment.clean_frameshifts</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">clean_frameshifts(aligned_ref::LongDNA{4},aligned_seq::LongDNA{4}; verbose::Bool=false)</code></pre><p>Takes a pairwise alignment of a reference (with known reading frame) and a sequence, and removes frameshift mutations  which don&#39;t respect the reference&#39;s reading frame. This is done by removing insertions from the alignment or inserting  ambigious nucleotides into deletions.</p><p><strong>Examples:</strong></p><ol><li>insertion</li></ol><p>ref: ATG-AACGTA  -&gt; cleaned<em>ref: ATGAACGTA  seq: ATGTAACGTA  -&gt; cleaned</em>seq: ATGAACGTA</p><ol><li>deletion</li></ol><p>ref: ATGAACGTA  -&gt; cleaned<em>ref: ATGAACGTA seq: ATG-ACGTA  -&gt; cleaned</em>seq: ATGNACGTA</p><p><strong>NOTE</strong> We always assume the readingFrame is 1</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/MurrellGroup/SeededAlignment.jl/blob/38a40c0887ec27f271ae16cb2f51d79048aec487/src/seq_alignment/clean_frameshifts.jl#L2-L19">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SeededAlignment.msa_codon_align-Tuple{BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}, Vector{BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}}}" href="#SeededAlignment.msa_codon_align-Tuple{BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}, Vector{BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}}}"><code>SeededAlignment.msa_codon_align</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">msa_codon_align(ref::LongDNA{4}, seqs::Vector{LongDNA{4}}; moveset::Moveset=std_codon_moveset(), scoring::ScoringScheme=std_scoring(), match_codons=true::Bool, use_seeded=true::Bool)</code></pre><p>Produces a fast codon aligment with consistent readingFrame. </p><p><strong>Arguments</strong></p><ul><li><code>ref::LongDNA{4}</code>: The reference DNA sequence, assumed to have an readingFrame which starts at its first nucleotide.</li><li><code>seqs::Vector{LongDNA{4}}</code>: A vector of DNA sequences to align against the reference.</li><li><code>moveset::Moveset</code>: Defines the allowable alignment operations (e.g. match nucleotide, single indel, readingFrame respecting triple indel). Assumes moveset is choosen so some moves respect the readingFrame.</li><li><code>score_params::ScoringScheme</code>: Scoring parameters for codon-aware alignment, including match/mismatch and gap penalties.</li></ul><p><strong>Returns</strong></p><ul><li><code>Vector{LongDNA{4}}</code>: A vector of aligned sequences (including the reference), with codon-aware gaps that preserve reading frame consistency.</li></ul><p><strong>Description</strong></p><p>Produces a fast multiple codon alignment based on a reference with known readingFrame, a moveset and a ScoringScheme.  This is done by computing a pairwise alignment with respect to the reference for each sequence and then cleaning up single indel noise.</p><p><strong>Example</strong></p><pre><code class="language-julia hljs">ref = LongDNA{4}(&quot;ATGACGTGA&quot;)  # Reference with known reading frame
seqs = [LongDNA{4}(&quot;ATGTCGTGA&quot;), LongDNA{4}(&quot;ATGACGAGA&quot;)]
moveset = std_codon_moveset()
score_params = std_scoring()

alignment = msa_codon_align(ref, seqs, moveset, score_params)</code></pre><p>NOTE: The sequences (ungapped) might not be preserved by the alignment by the cleanup step. Some nucleotides might be inserted or removed.  </p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/MurrellGroup/SeededAlignment.jl/blob/38a40c0887ec27f271ae16cb2f51d79048aec487/src/seq_alignment/msa_codon_align.jl#L1-L29">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SeededAlignment.nw_align-Tuple{BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}, BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}}" href="#SeededAlignment.nw_align-Tuple{BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}, BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}}"><code>SeededAlignment.nw_align</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">nw_align(A::LongDNA{4},B::LongDNA{4}; moveset::Moveset=std_noisy_moveset(), scoring::ScoringScheme=std_scoring()</code></pre><p>Needleman_Wunsch wrapper - no reference, i.e. makes no assumptions about the two sequences. </p><p>Computes an optimal global pairwise alignment of two ungapped <code>LongDNA{4}</code> sequence   with respect to the <code>Moveset</code> and the <code>ScoringScheme</code>. </p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/MurrellGroup/SeededAlignment.jl/blob/38a40c0887ec27f271ae16cb2f51d79048aec487/src/seq_alignment/needleman_wunsch.jl#L1-L11">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SeededAlignment.nw_align-Tuple{}" href="#SeededAlignment.nw_align-Tuple{}"><code>SeededAlignment.nw_align</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">nw_align(; ref::LongDNA{4}, query::LongDNA{4}, moveset::Moveset=std_codon_moveset(), scoring::ScoringScheme=std_scoring(),
    do_clean_frameshifts=false::Bool, verbose=false::Bool, match_codons=true::Bool)</code></pre><p>Needleman_Wunsch wrapper - reference informed, i.e. assumes one of the sequence has intact reading frame. </p><p>Optimally aligns a <code>query</code> sequence to a <code>ref</code> sequence using a codon-aware <code>Moveset</code> and <code>ScoringScheme</code>. In particular ...  ... </p><p><strong>NOTE</strong> We always assume the readingFrame is 1</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/MurrellGroup/SeededAlignment.jl/blob/38a40c0887ec27f271ae16cb2f51d79048aec487/src/seq_alignment/needleman_wunsch.jl#L25-L37">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SeededAlignment.read_fasta-Tuple{String}" href="#SeededAlignment.read_fasta-Tuple{String}"><code>SeededAlignment.read_fasta</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">read_fasta(filepath::String)</code></pre><p>Reads in a fasta file and returns a tuple of (seqnames, seqs).</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/MurrellGroup/SeededAlignment.jl/blob/38a40c0887ec27f271ae16cb2f51d79048aec487/src/utils/fasta_io.jl#L1-L5">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SeededAlignment.seed_chain_align-Tuple{BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}, BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}}" href="#SeededAlignment.seed_chain_align-Tuple{BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}, BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}}"><code>SeededAlignment.seed_chain_align</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">seed_chain_align(A::LongDNA{4}, B::LongDNA{4}, moveset::Moveset, ScoringScheme::ScoringScheme)</code></pre><p>Perform pairwise alignment between two DNA sequences using a seeding strategy.</p><p><strong>Arguments</strong></p><ul><li><code>A::LongDNA{4}</code>: The first DNA sequence to align.</li><li><code>B::LongDNA{4}</code>: The second DNA sequence to align.</li><li><code>moveset::Moveset</code>: Defines allowed alignment moves (e.g., match, mismatch, gap).</li><li><code>ScoringScheme::ScoringScheme</code>: Scoring scheme for e.g. matches, mismatches, and gaps.</li></ul><p><strong>Returns</strong></p><ul><li><code>Vector{LongDNA{4}}</code>: A vector the two aligned sequences.</li></ul><p><strong>Description</strong></p><p>This function performs a fast (sub-quadratic) pairwise alignment using seeding and chaining. It identifies high-scoring seed matches between <code>A</code> and <code>B</code>, chains them to build a candidate alignment skeleton, and fills gaps between seeds using dynamic programming guided by <code>moveset</code> and <code>ScoringScheme</code>.</p><p><strong>Example</strong></p><pre><code class="language-julia hljs">A = LongDNA{4}(&quot;ACGTACGT&quot;)
B = LongDNA{4}(&quot;ACGTTGCA&quot;)
moveset = std_codon_moveset()
ScoringScheme = std_scoring()

alignment = seed_chain_align(A, B, moveset, ScoringScheme)</code></pre><p><strong>seed<em>chain</em>align wrapper - default noisy i.e no reference sequence</strong></p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/MurrellGroup/SeededAlignment.jl/blob/38a40c0887ec27f271ae16cb2f51d79048aec487/src/seq_alignment/seed_chain_align.jl#L1-L28">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SeededAlignment.std_codon_moveset-Tuple{}" href="#SeededAlignment.std_codon_moveset-Tuple{}"><code>SeededAlignment.std_codon_moveset</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">std_codon_moveset()</code></pre><p>Return a standard codon-aware <code>Moveset</code> for sequence alignment.</p><p><strong>Returns</strong></p><ul><li><code>Moveset</code>: Contains codon-aware match and gap moves.</li></ul><p><strong>Moves</strong></p><ul><li>Match: <code>Move(1, 0.0)</code></li><li>Horizontal (gaps in reference sequence):<ul><li><code>Move(1, 2.0, 1, 0, 1, 0, false)</code></li><li><code>Move(3, 2.0, 1, 0, 3, 0, true)</code></li></ul></li><li>Vertical (gaps in query sequence):<ul><li><code>Move(1, 2.0, 1, 0, 1, 0, false)</code></li><li><code>Move(3, 2.0, 1, 0, 3, 0, true)</code></li></ul></li></ul><p>Use this as a default <code>Moveset</code> for codon-preserving alignments.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/MurrellGroup/SeededAlignment.jl/blob/38a40c0887ec27f271ae16cb2f51d79048aec487/src/alignmentOptions/moveset.jl#L111-L129">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SeededAlignment.std_noisy_moveset-Tuple{}" href="#SeededAlignment.std_noisy_moveset-Tuple{}"><code>SeededAlignment.std_noisy_moveset</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">pairwise_noisy_moveset()</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/MurrellGroup/SeededAlignment.jl/blob/38a40c0887ec27f271ae16cb2f51d79048aec487/src/alignmentOptions/moveset.jl#L136-L139">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SeededAlignment.std_scoring-Tuple{}" href="#SeededAlignment.std_scoring-Tuple{}"><code>SeededAlignment.std_scoring</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">std_codon_scoring()</code></pre><p>Return a standard codon-aware <code>ScoringScheme</code> for alignment.</p><p><strong>Returns</strong></p><ul><li><code>ScoringScheme</code>: Default scoring parameters for codon alignment.</li></ul><p><strong>Parameters</strong></p><ul><li>Match score: <code>0.0</code></li><li>Mismatch score: <code>0.3</code></li><li>Gap extension score: <code>0.1</code></li><li>Allow extension at sequence ends: <code>true</code> (both ends)</li><li>K-mer length for seeding: <code>18</code></li></ul><p>Use this as a default <code>ScoringScheme</code> for codon-preserving alignments.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/MurrellGroup/SeededAlignment.jl/blob/38a40c0887ec27f271ae16cb2f51d79048aec487/src/alignmentOptions/scoring_scheme.jl#L72-L88">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SeededAlignment.write_fasta-Tuple{String, Tuple{BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}, BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}}}" href="#SeededAlignment.write_fasta-Tuple{String, Tuple{BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}, BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}}}"><code>SeededAlignment.write_fasta</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">write_fasta(filepath::String, sequences::Tuple{LongDNA{4},LongDNA{4}}; seq_names = nothing)</code></pre><p>Writes a fasta file from a Tuple of sequences, with optional seq_names.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/MurrellGroup/SeededAlignment.jl/blob/38a40c0887ec27f271ae16cb2f51d79048aec487/src/utils/fasta_io.jl#L31-L35">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SeededAlignment.write_fasta-Tuple{String, Vector{BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}}}" href="#SeededAlignment.write_fasta-Tuple{String, Vector{BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}}}"><code>SeededAlignment.write_fasta</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">write_fasta(filepath::String, sequences::Vector{LongDNA{4}}; seq_names = nothing)</code></pre><p>Writes a fasta file from a vector of sequences, with optional seq_names.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/MurrellGroup/SeededAlignment.jl/blob/38a40c0887ec27f271ae16cb2f51d79048aec487/src/utils/fasta_io.jl#L14-L18">source</a></section></article></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../">« Home</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.14.1 on <span class="colophon-date" title="Tuesday 15 July 2025 02:25">Tuesday 15 July 2025</span>. Using Julia version 1.11.6.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
