<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>API Reference · SeededAlignment.jl</title><meta name="title" content="API Reference · SeededAlignment.jl"/><meta property="og:title" content="API Reference · SeededAlignment.jl"/><meta property="twitter:title" content="API Reference · SeededAlignment.jl"/><meta name="description" content="Documentation for SeededAlignment.jl."/><meta property="og:description" content="Documentation for SeededAlignment.jl."/><meta property="twitter:description" content="Documentation for SeededAlignment.jl."/><meta property="og:url" content="https://MurrellGroup.github.io/SeededAlignment.jl/api/"/><meta property="twitter:url" content="https://MurrellGroup.github.io/SeededAlignment.jl/api/"/><link rel="canonical" href="https://MurrellGroup.github.io/SeededAlignment.jl/api/"/><script data-outdated-warner src="../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../search_index.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../">SeededAlignment.jl</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../">Home</a></li><li class="is-active"><a class="tocitem" href>API Reference</a><ul class="internal"><li><a class="tocitem" href="#Core-Methods"><span>Core Methods</span></a></li><li><a class="tocitem" href="#Utilities"><span>Utilities</span></a></li><li><a class="tocitem" href="#Core-Types"><span>Core Types</span></a></li><li><a class="tocitem" href="#Standard-Parameters"><span>Standard Parameters</span></a></li><li><a class="tocitem" href="#Index"><span>Index</span></a></li><li><a class="tocitem" href="#Index-Docstrings"><span>Index Docstrings</span></a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>API Reference</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>API Reference</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/MurrellGroup/SeededAlignment.jl" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/MurrellGroup/SeededAlignment.jl/blob/main/docs/src/api.md" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="API-Reference"><a class="docs-heading-anchor" href="#API-Reference">API Reference</a><a id="API-Reference-1"></a><a class="docs-heading-anchor-permalink" href="#API-Reference" title="Permalink"></a></h1><p>This section provides detailed documentation for all exported types and functions in <code>SeededAlignment.jl</code>.</p><h2 id="Core-Methods"><a class="docs-heading-anchor" href="#Core-Methods">Core Methods</a><a id="Core-Methods-1"></a><a class="docs-heading-anchor-permalink" href="#Core-Methods" title="Permalink"></a></h2><ul><li><a href="#SeededAlignment.seed_chain_align-Tuple{BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}, BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}}"><code>seed_chain_align</code></a></li><li><a href="#SeededAlignment.msa_codon_align-Tuple{BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}, Vector{BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}}}"><code>msa_codon_align</code></a> </li><li><a href="#SeededAlignment.clean_frameshifts-Tuple{BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}, BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}}"><code>clean_frameshifts</code></a></li><li><a href="#SeededAlignment.nw_align-Tuple{BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}, BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}}"><code>nw_align</code></a></li></ul><h2 id="Utilities"><a class="docs-heading-anchor" href="#Utilities">Utilities</a><a id="Utilities-1"></a><a class="docs-heading-anchor-permalink" href="#Utilities" title="Permalink"></a></h2><p>methods for reading and writing of sequence data in FASTA format.</p><ul><li><a href="#SeededAlignment.read_fasta-Tuple{String}"><code>read_fasta</code></a></li><li><a href="#SeededAlignment.write_fasta-Tuple{String, Vector{BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}}}"><code>write_fasta</code></a></li></ul><h2 id="Core-Types"><a class="docs-heading-anchor" href="#Core-Types">Core Types</a><a id="Core-Types-1"></a><a class="docs-heading-anchor-permalink" href="#Core-Types" title="Permalink"></a></h2><ul><li><a href="#SeededAlignment.Move"><code>Move</code></a></li><li><a href="#SeededAlignment.Moveset"><code>Moveset</code></a></li><li><a href="#SeededAlignment.ScoringScheme"><code>ScoringScheme</code></a></li><li>[<code>LongDNA{4}</code>] - Array specialized for DNA from BioSequences.jl</li></ul><h2 id="Standard-Parameters"><a class="docs-heading-anchor" href="#Standard-Parameters">Standard Parameters</a><a id="Standard-Parameters-1"></a><a class="docs-heading-anchor-permalink" href="#Standard-Parameters" title="Permalink"></a></h2><ul><li><a href="#SeededAlignment.STD_SCORING"><code>STD_SCORING</code></a></li><li><a href="#SeededAlignment.STD_CODON_MOVESET"><code>STD_CODON_MOVESET</code></a></li><li><a href="#SeededAlignment.STD_NOISY_MOVESET"><code>STD_NOISY_MOVESET</code></a></li></ul><h2 id="Index"><a class="docs-heading-anchor" href="#Index">Index</a><a id="Index-1"></a><a class="docs-heading-anchor-permalink" href="#Index" title="Permalink"></a></h2><p>A complete alphabetical listing of all documented functions.</p><ul><li><a href="#SeededAlignment.STD_CODON_MOVESET"><code>SeededAlignment.STD_CODON_MOVESET</code></a></li><li><a href="#SeededAlignment.STD_NOISY_MOVESET"><code>SeededAlignment.STD_NOISY_MOVESET</code></a></li><li><a href="#SeededAlignment.STD_SCORING"><code>SeededAlignment.STD_SCORING</code></a></li><li><a href="#SeededAlignment.Move"><code>SeededAlignment.Move</code></a></li><li><a href="#SeededAlignment.Moveset"><code>SeededAlignment.Moveset</code></a></li><li><a href="#SeededAlignment.ScoringScheme-Tuple{}"><code>SeededAlignment.ScoringScheme</code></a></li><li><a href="#SeededAlignment.ScoringScheme"><code>SeededAlignment.ScoringScheme</code></a></li><li><a href="#SeededAlignment.clean_frameshifts-Tuple{BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}, BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}}"><code>SeededAlignment.clean_frameshifts</code></a></li><li><a href="#SeededAlignment.clean_frameshifts-Tuple{BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}, Vector{BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}}}"><code>SeededAlignment.clean_frameshifts</code></a></li><li><a href="#SeededAlignment.msa_codon_align-Tuple{BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}, Vector{BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}}}"><code>SeededAlignment.msa_codon_align</code></a></li><li><a href="#SeededAlignment.nw_align-Tuple{}"><code>SeededAlignment.nw_align</code></a></li><li><a href="#SeededAlignment.nw_align-Tuple{BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}, BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}}"><code>SeededAlignment.nw_align</code></a></li><li><a href="#SeededAlignment.read_fasta-Tuple{String}"><code>SeededAlignment.read_fasta</code></a></li><li><a href="#SeededAlignment.seed_chain_align-Tuple{}"><code>SeededAlignment.seed_chain_align</code></a></li><li><a href="#SeededAlignment.seed_chain_align-Tuple{BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}, BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}}"><code>SeededAlignment.seed_chain_align</code></a></li><li><a href="#SeededAlignment.write_fasta-Union{Tuple{N}, Tuple{String, NTuple{N, BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}}}} where N"><code>SeededAlignment.write_fasta</code></a></li><li><a href="#SeededAlignment.write_fasta-Tuple{String, Vector{BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}}}"><code>SeededAlignment.write_fasta</code></a></li></ul><h2 id="Index-Docstrings"><a class="docs-heading-anchor" href="#Index-Docstrings">Index Docstrings</a><a id="Index-Docstrings-1"></a><a class="docs-heading-anchor-permalink" href="#Index-Docstrings" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SeededAlignment.Move" href="#SeededAlignment.Move"><code>SeededAlignment.Move</code></a> — <span class="docstring-category">Type</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">Move(ref::Bool, step_length::Int64, score::Float64, extendable::Float64)</code></pre><p>Represents a gap move during alignment. The <code>Move</code> instance represents either an insertion or deletion depending on how it is passed to the <code>Moveset</code> instance - collection of <code>Move</code> instances used in alignment methods. </p><p>For example, if Move.step_length = 3 then that represents a gap of length 3 (either insertion or deletion). </p><p><strong>Extended Help</strong></p><p><strong>Fields</strong></p><p>-<code>ref::Bool</code>: Whether move respects the coding reading frame -<code>step_length::Int64</code>: Length of gap in alignment (must be  1,2 or 3) -<code>score::Float64</code>: penalty for using the move - cost of adding the gaps to the alignment (must be &lt; 0) -<code>extendable::Bool</code>: Whether the move can be affinely extended by another move. </p><p><strong>Constructors</strong></p><p>-<code>Move(; ref=false::Bool, step_length::Int64, score::Float64, extendable=false)</code>: keyword constructor for more easily constructing Moves. Has some default values but requires at least <code>step_length</code> and <code>score</code> to be provided. </p><p>-<code>RefMove(; score::Float64)</code>:  Constructor for a <code>Move</code> that respect coding reading frame. Only requires <code>score</code> argument.  Returns: <code>Move(ref=true, step_length=3, score=score, extendable=true)</code></p><p>-<code>FrameshiftMove(; step_length::Int64, score::Float64, extendable::Bool=false)</code>:  Constructor for a <code>Move</code> that cause frameshifts and breaks reading frame symmetry. Requires arguments <code>step_length</code>, <code>score</code>, and <code>extendable</code>. Returns: <code>Move(ref=false, step_length=step_length, score=score, extendable=extendable)</code></p><p><strong>Examples</strong></p><ol><li>codon moveset with no frameshifts</li></ol><pre><code class="language-julia hljs"># represents codon insertion or codon deletion
codon_indel = RefMove(score=1.0)
#= passing to moveset solidifies what the allowed alignment operations are,
namely single codon insertions and deletions. 
=#
Moveset(ref_insertions = (codon_indel,), ref_deletions = (codon_indel,))</code></pre><ol><li>codon moveset with frameshift moves allowed</li></ol><pre><code class="language-julia hljs"># represents codon insertion or codon deletion
codon_indel = RefMove(score=-1.0)
# represents frameshift causing insertion or deletion
frm_indel = FrameshiftMove(score=-1.5, step_length=1, extendable=true)
#= passing to moveset solidifies what the allowed alignment operations are,
namely single codon insertions and deletions and single nucleotide indels that are extendable. 
=#
Moveset(ref_insertions = (codon_indel,frm_indel), ref_deletions = (codon_indel, frm_indel))</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/MurrellGroup/SeededAlignment.jl/blob/429af62de75af9f9599992e8f4babe1c63d70471/src/alignmentOptions/moveset.jl#L2-L57">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SeededAlignment.Moveset" href="#SeededAlignment.Moveset"><code>SeededAlignment.Moveset</code></a> — <span class="docstring-category">Type</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">Moveset{X,Y}(; ref_insertions::NTuple{X,Move}, ref_deletions::NTuple{Y,Move})</code></pre><p>Represents a collection of <code>Move</code> instances that are either insertions or deletions</p><p><strong>Extended Help</strong></p><p><strong>Fields</strong></p><p>-<code>vert_moves::NTuple{X,Move}</code>: insertions relative to reference - gap operation in top/first provided sequence  -<code>hor_moves::NTuple{Y,Move}</code>: deletions relative to reference - gap operations in left/second provided sequence </p><p><strong>Examples</strong></p><ol><li>codon moveset with no frameshifts</li></ol><pre><code class="language-julia hljs"># represents codon insertion or codon deletion
codon_indel = RefMove(score=1.0)
#= passing to moveset solidifies what the allowed alignment operations are,
namely single codon insertions and deletions. 
=#
ms = Moveset(ref_insertions = (codon_indel,), ref_deletions = (codon_indel,))</code></pre><ol><li>codon moveset with frameshift moves allowed</li></ol><pre><code class="language-julia hljs"># represents codon insertion or codon deletion
codon_indel = RefMove(score=-1.0)
# represents frameshift causing insertion or deletion
frm_indel = FrameshiftMove(score=-1.5, step_length=1, extendable=true)
#= passing to moveset solidifies what the allowed alignment operations are,
namely single codon insertions and deletions and single nucleotide indels that are extendable. 
=#
ms = Moveset(ref_insertions = (codon_indel,frm_indel), ref_deletions = (codon_indel, frm_indel))</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/MurrellGroup/SeededAlignment.jl/blob/429af62de75af9f9599992e8f4babe1c63d70471/src/alignmentOptions/moveset.jl#L85-L122">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SeededAlignment.ScoringScheme" href="#SeededAlignment.ScoringScheme"><code>SeededAlignment.ScoringScheme</code></a> — <span class="docstring-category">Type</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">ScoringScheme(;
	extension_score::Float64=-0.3,
	kmer_length::Int64=18,
	edge_ext_begin::Bool=true,
	edge_ext_end::Bool=true,
	nucleotide_match_score::Float64 = 0.0,
	nucleotide_mismatch_score::Float64 = -0.8,
	nucleotide_score_matrix::Union{Nothing,Matrix{Float64}} = nothing
	codon_match_bonus_score::Float64 = 6.0</code></pre><p>)</p><p><code>ScoringScheme</code> defines how matches, mismatches, and partially how  gaps are scored during sequence alignment. Beyond alignment operations governed by <code>Moveset</code> it encompasses everything else we can customize about the alignment process. </p><p>This struct is typically passed to functions like <code>seed_chain_align</code>, <code>nw_align</code> and <code>msa_codon_align</code>.</p><p><strong>Fields</strong></p><p>-<code>extension_score::Float64=-0.3</code>: penalty for affinely extending a gap that is already open.  -<code>kmer_length::Int64=18</code>: length of seeds used in seed<em>chain</em>align -<code>edge_ext_begin=true</code>: Whether to subsidize gaps in begining of the alignment -<code>edge_ext_end=true</code>: Whether to subsidize gaps in end of the alignment -<code>nucleotide_match_score = 0.0</code>: score awarded for matching nucleotides (has to be &gt;= 0) -<code>nucleotide_mismatch_score = -0.8</code>: penalty for matching distinct nucleotides (has to be &lt; 0) -<code>nucleotide_score_matrix::Union{Nothing, Matrix{Float64}}</code>: optional custom matrix for scoring nucleotide substitutions. -<code>codon_match_bonus_score = 6.0</code>: score awarded for matching codons</p><p><strong>example</strong></p><pre><code class="language-julia hljs">score_params = ScoringScheme(extension_score = -0.2, mismatch_score = -0.7) # (everything else will be keept at default values)
A = LongDNA{4}(&quot;ATGATGATAA&quot;)
B = LongDNA{4}(&quot;ATGACCCGATAA&quot;)
seed_chain_align(A,B scoring=score_params)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/MurrellGroup/SeededAlignment.jl/blob/429af62de75af9f9599992e8f4babe1c63d70471/src/alignmentOptions/scoring_scheme.jl#L1-L35">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SeededAlignment.ScoringScheme-Tuple{}" href="#SeededAlignment.ScoringScheme-Tuple{}"><code>SeededAlignment.ScoringScheme</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">ScoringScheme(;
	extension_score::Float64=-0.3,
	kmer_length::Int64=12,
	edge_ext_begin::Bool=true,
	edge_ext_end::Bool=true,
	nucleotide_match_score::Float64 = 0.0,
	nucleotide_mismatch_score::Float64 = -0.4,
	nucleotide_score_matrix::Union{Nothing,Matrix{Float64}} = nothing
	codon_match_bonus_score::Float64 = 6.0</code></pre><p>)</p><p>keyword constructor for <code>ScoringScheme</code> with helpful default parameters. It is important to note that supplying a custom matrix for <code>nucleotide_score_matrix</code> might slow performance of some alignment methods.  </p><p><strong>Extended Help</strong></p><p>Slow down might occur due to allocating <code>ScoreScheme</code> to the heap instead of the stack since <code>matrix{Float64}</code> is dynamically sized. </p><p><strong>Arguments</strong></p><p>-<code>extension_score::Float64=-0.3</code>: penalty for affinely extending a gap that is already open.  -<code>kmer_length::Int64=18</code>: length of seeds used in seed<em>chain</em>align -<code>edge_ext_begin=true</code>: Whether to subsidize gaps in begining of the alignment -<code>edge_ext_end=true</code>: Whether to subsidize gaps in end of the alignment -<code>nucleotide_match_score = 0.0</code>: score awarded for matching nucleotides (has to be &gt;= 0) -<code>nucleotide_mismatch_score = -0.8</code>: penalty for matching distinct nucleotides (has to be &lt; 0) -<code>nucleotide_score_matrix::Union{Nothing, Matrix{Float64}}</code>: optional custom matrix for scoring nucleotide substitutions. -<code>codon_match_bonus_score = 6.0</code>: score awarded for matching codons</p><p><strong>Returns</strong></p><p>-<code>ScoringScheme</code></p><p><strong>example</strong></p><p>```julia score<em>params = ScoringScheme(extension</em>score = -0.2, mismatch<em>score = -0.7) # (everything else will be keept at default values) A = LongDNA{4}(&quot;ATGATGATAA&quot;) B = LongDNA{4}(&quot;ATGACCCGATAA&quot;) seed</em>chain<em>align(A,B scoring=score</em>params)</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/MurrellGroup/SeededAlignment.jl/blob/429af62de75af9f9599992e8f4babe1c63d70471/src/alignmentOptions/scoring_scheme.jl#L74-L114">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SeededAlignment.clean_frameshifts-Tuple{BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}, BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}}" href="#SeededAlignment.clean_frameshifts-Tuple{BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}, BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}}"><code>SeededAlignment.clean_frameshifts</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">clean_frameshifts(aligned_ref::LongDNA{4}, aligned_seq::LongDNA{4}; verbose::Bool=false)</code></pre><p>(clean_frameshifts - cleans pairwise alignments of frameshifts)</p><p>Takes a global pairwise alignment of a CDS anchor/reference <code>aligned_ref</code> and a CDS that may contain frameshift errors <code>aligned_seq</code>, and removes frameshift mutations  which don&#39;t respect the reference&#39;s reading frame in the alignment. This is done by removing insertions from the alignment or inserting  ambigious nucleotides into deletions.</p><p><strong>Extended Help</strong></p><p><strong>Arguments</strong></p><ul><li><code>aligned_ref::LongDNA{4}</code>: aligned anchored trusted CDS which decides the reading frame coordinates in the alignment</li><li><code>aligned_seq::LongDNA{4}</code>: aligned CDS (with possible frameshifts due to e.g. sequencing or annotation errors) which is aligned to <code>ref</code> and adopts its reading frame coordinates</li><li><code>verbose::Bool</code>: Whether to verbosely display what edits were made during the cleaning of frameshifts</li></ul><p><strong>Returns</strong></p><ul><li><code>Tuple{LongDNA{4},LongDNA{4}}</code>: Cleaned global pairwise alignment that represents a protein alignment on a nucleotide level </li></ul><p><strong>Examples</strong></p><pre><code class="nohighlight hljs">1. insertion
ref: ATG-AACGTA  -&gt; cleaned_ref: ATGAACGTA 
seq: ATGTAACGTA  -&gt; cleaned_seq: ATGAACGTA

2. deletion
ref: ATGAACGTA  -&gt; cleaned_ref: ATGAACGTA
seq: ATG-ACGTA  -&gt; cleaned_seq: ATGNACGTA

3. longer deletion
ref: ATGAACGTA  -&gt; cleaned_ref: ATGAACGTA
seq: AT-----TA  -&gt; cleaned_seq: ATN---NTA</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/MurrellGroup/SeededAlignment.jl/blob/429af62de75af9f9599992e8f4babe1c63d70471/src/seq_alignment/clean_frameshifts.jl#L2-L35">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SeededAlignment.clean_frameshifts-Tuple{BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}, Vector{BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}}}" href="#SeededAlignment.clean_frameshifts-Tuple{BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}, Vector{BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}}}"><code>SeededAlignment.clean_frameshifts</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">clean_frameshifts(aligned_ref::LongDNA{4}, aligned_seqs::Vector{LongDNA{4}})</code></pre><p>(clean_frameshifts - cleans multiple sequence alignments of frameshifts)</p><p>Clean a multiple sequence alignment provided one of the sequence is a reference sequence. This is done by projecting the multiple sequence alignment to  a collection of pairwise alignments relative to the reference <code>aligned_ref</code>; cleaning those and then scaffolding the results to recover a cleaned multiple sequence alignment.</p><p><strong>Extended Help</strong></p><p><strong>Arguments</strong></p><ul><li><code>aligned_ref::LongDNA{4}</code>: aligned anchored trusted CDS which decides the reading frame coordinates in the alignment</li><li><code>aligned_seqs::LongDNA{4}</code>: aligned coding sequences (with possible frameshifts due to e.g. sequencing or annotation errors) which are aligned to <code>ref</code> and adopts its reading frame coordinates</li><li><code>verbose::Bool</code>: Whether to verbosely display what edits were made during the cleaning of frameshifts</li></ul><p><strong>Returns</strong></p><ul><li><code>cleaned_msa::Vector{LongDNA{4}}</code>: a frameshift-free multiple sequence alignment</li></ul><p><strong>Example</strong></p><pre><code class="language-julia hljs">aligned_seqs = Vector{LongDNA{4}}(undef, 4)
# original unclean msa
            aligned_ref =     LongDNA{4}(&quot;ATG---TTTCCCGGGT-AA&quot;)
            aligned_seqs[1] = LongDNA{4}(&quot;-TG------CCCGGGT-A-&quot;)
            aligned_seqs[2] = LongDNA{4}(&quot;ATGAAATTTCCCGGGT-AA&quot;)
            aligned_seqs[3] = LongDNA{4}(&quot;ATGAAA----CCGGGT-AA&quot;)
            aligned_seqs[4] = LongDNA{4}(&quot;ATG---TTTCCCGGGTTAA&quot;)
# produce clean msa
cleaned_msa = clean_frameshifts(aligned_ref, aligned_seqs)
#= results
            cleaned_msa[1] = LongDNA{4}(&quot;ATG---TTTCCCGGGTAA&quot;) # ref sequence
            cleaned_msa[2] = LongDNA{4}(&quot;NTG------CCCGGGTAN&quot;)
            cleaned_msa[3] = LongDNA{4}(&quot;ATGAAATTTCCCGGGTAA&quot;)
            cleaned_msa[4] = LongDNA{4}(&quot;ATGAAA---NCCGGGTAA&quot;)
            cleaned_msa[5] = LongDNA{4}(&quot;ATG---TTTCCCGGGTAA&quot;)

=#</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/MurrellGroup/SeededAlignment.jl/blob/429af62de75af9f9599992e8f4babe1c63d70471/src/seq_alignment/clean_frameshifts.jl#L202-L242">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SeededAlignment.msa_codon_align-Tuple{BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}, Vector{BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}}}" href="#SeededAlignment.msa_codon_align-Tuple{BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}, Vector{BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}}}"><code>SeededAlignment.msa_codon_align</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">msa_codon_align(ref::LongDNA{4}, seqs::Vector{LongDNA{4}}; moveset::Moveset=STD_CODON_MOVESET, scoring::ScoringScheme=STD_SCORING, codon_scoring_on=true::Bool)</code></pre><p>Computes a visual global MSA (multiple Sequence alignment) of coding sequences based on pairwise alignments to a trusted CDS reference <code>ref</code> used as an anchor to  determine the apprioate reading frame coordinates for the other coding sequences (which may contain frameshift errors). Possible frameshifts errors in the  pairwise alignments are cleaned up and then scaffolded to create a multiple sequence alignment. This is done by left-stacking codon insertions relative to  the reference. </p><p>Note that this doesn&#39;t qualify as a proper multiple sequence alignment in the traditional sense since the aligned sequences are only scored as being  aligned against the reference sequnece and not each other. </p><p>Even so, it can still provides a useful visualization or approximation for a protein multiple sequence alignment on a nucleotide level. </p><p><strong>Extended Help</strong></p><p><strong>Arguments</strong></p><ul><li><code>ref::LongDNA{4}</code>: Anchored trusted CDS which decides the reading frame coordinates in the alignment</li><li><code>seqs::Vector{LongDNA{4}}</code>: Coding Sequences (with possible frameshifts due to e.g. sequencing or annotation errors) which are aligned to <code>ref</code> and adopts its reading frame coordinates. </li><li><code>moveset::Moveset=STD_CODON_MOVESET</code>: Defines allowable alignment moves (e.g., codon insertions/deletions).</li><li><code>scoring::ScoringScheme=STD_SCORING</code>: The scoring scheme used during alignment.</li><li><code>codon_scoring_on::Bool=true</code>: Whether to apply additional scoring on codon-level </li></ul><p><strong>Returns</strong></p><ul><li><code>msa::Vector{LongDNA{4}}</code>: a frameshift-free multiple sequence alignment</li></ul><p><strong>Example</strong></p><ol><li>codon alignment with no frameshifts present in inputs </li></ol><pre><code class="language-julia hljs">ref =  LongDNA{4}(&quot;ATGTTTCCCGGGTAA&quot;)
seq1 = LongDNA{4}(&quot;ATGTTTTTTCCCGGGTAA&quot;)
seq2 = LongDNA{4}(&quot;ATGATGTTTTTTCCCGGGTAAGGG&quot;)
seq3 = LongDNA{4}(&quot;ATGCCCGGG&quot;)

msa = msa_codon_align(ref, [seq1,seq2,seq3])
#= alignment results:

msa[1] == LongDNA{4}(&quot;---ATG---TTTCCCGGGTAA---&quot;)
msa[2] == LongDNA{4}(&quot;---ATGTTTTTTCCCGGGTAA---&quot;)
msa[3] == LongDNA{4}(&quot;ATGATGTTTTTTCCCGGGTAAGGG&quot;)
msa[4] == LongDNA{4}(&quot;---ATG------CCCGGG------&quot;)

=#</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/MurrellGroup/SeededAlignment.jl/blob/429af62de75af9f9599992e8f4babe1c63d70471/src/seq_alignment/msa_codon_align.jl#L1-L47">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SeededAlignment.nw_align-Tuple{BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}, BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}}" href="#SeededAlignment.nw_align-Tuple{BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}, BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}}"><code>SeededAlignment.nw_align</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">nw_align(A::LongDNA{4}, B::LongDNA{4}; moveset::Moveset=STD_NOISY_MOVESET, scoring::ScoringScheme=STD_SCORING)</code></pre><p>(Needleman-wunsch wrapper - DE-NOVO)</p><p>Computes an optimal global pairwise alignment of the two ungapped DNA sequences <code>A</code> and <code>B</code>. This is done purely semantically without any awareness of protein encoding.</p><p><strong>Extended Help</strong></p><p><strong>Arguments</strong></p><ul><li><code>A::LongDNA{4}</code>: 1st DNA sequence to be aligned</li><li><code>B::LongDNA{4}</code>: 2nd DNA sequence to be aligned</li><li><code>moveset::Moveset=STD_NOISY_MOVESET</code>: Defines allowable alignment moves (e.g. insertions/deletions and their penalty)</li><li><code>scoring::ScoreScheme=STD_SCORING</code>: Defines alignment scoring together with moveset</li></ul><p><strong>Returns</strong></p><ul><li><code>Tuple{LongDNA{4},LongDNA{4}}</code>: Tuple representation of pairwise alignment of DNA sequences <code>A</code> and <code>B</code>.</li></ul><p><strong>Examples</strong></p><pre><code class="language-julia hljs">A = LongDNA{4}(&quot;AATGCTC&quot;)
B = LongDNA{4}(&quot;ACATGTC&quot;)
# produce alignment 
alignment = nw_align(A, B)
println(alignment)
#= resulting alignment:

alignment = (
	LongDNA{4}(&quot;A-ATGCTC&quot;), 
	LongDNA{4}(&quot;ACATG-TC&quot;)
)

=#</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/MurrellGroup/SeededAlignment.jl/blob/429af62de75af9f9599992e8f4babe1c63d70471/src/seq_alignment/needleman_wunsch.jl#L1-L38">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SeededAlignment.nw_align-Tuple{}" href="#SeededAlignment.nw_align-Tuple{}"><code>SeededAlignment.nw_align</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">nw_align(; 
    	ref::LongDNA{4}, 
    	query::LongDNA{4}, 
    	moveset::Moveset = STD_CODON_MOVESET, 
    	scoring::ScoringScheme = STD_SCORING,
    	codon_scoring_on::Bool = false,
    	do_clean_frameshifts::Bool = true, 
    	verbose::Bool = false)</code></pre><p>(Needleman-wunsch wrapper - CODING given trusted CDS anchor/reference)</p><p>Produces an optimal global pairwise alignment of two ungapped CDS (Coding DNA Sequences) <code>ref</code> and <code>query</code> by using an <code>ref</code> as an anchor to determine the apprioate reading frame coordinates for <code>query</code>. </p><p><strong>Extended Help</strong></p><p><strong>Arguments</strong></p><ul><li><code>ref::LongDNA{4}</code>: Anchored trusted CDS which decides the reading frame coordinates in the alignment</li><li><code>query::LongDNA{4}</code>: CDS (with possible frameshifts due to e.g. sequencing or annotation errors) which is aligned to <code>ref</code> and adopts its reading frame coordinates. </li><li><code>moveset::Moveset = STD_CODON_MOVESET</code>: Defines allowable alignment moves (e.g. insertions/deletions and their penalty)</li><li><code>scoring::ScoringScheme = STD_SCORING</code>: Defines alignment scoring together with moveset</li><li><code>codon_scoring_on::Bool = false</code>: Whether to apply additional scoring on codon-level </li><li><code>do_clean_frameshifts::Bool = true</code>: Whether to clean the alignment output of gaps which cause frameshifts (IMPORTANT: produces a protein alignment on a nucleotide level)</li><li><code>verbose::Bool = false</code>: Whether to verbosely display what edits were made during the cleaning of frameshifts. </li></ul><p><strong>Returns</strong></p><ul><li><code>Tuple{LongDNA{4},LongDNA{4}}</code>: Tuple representation of pairwise alignment of DNA sequences <code>ref</code> and <code>query</code>. </li></ul><p>Note that this represents a protein alignment on a nucleotide level if <code>(do_clean_frameshifts == true)</code>.</p><p><strong>Examples</strong></p><pre><code class="language-julia hljs">anchor_CDS =    LongDNA{4}(&quot;ATGCCAGTA&quot;)
# untrusted_CDS may contain some frameshift errors due to e.g. sequencing or annotation errors.
untrusted_CDS = LongDNA{4}(&quot;ATGTA&quot;) 
# frameshift errors are removed from the cleaned alignment
cleaned_CDS_alignment = nw_align(ref=anchor_CDS, query=untrusted_CDS, clean_frameshifts=true)
println(cleaned_CDS_alignment)
#= resulting alignment:

cleaned_CDS_alignment = (
	LongDNA{4}(&quot;ATGCCAGTA&quot;), 
	LongDNA{4}(&quot;ATG---NTA&quot;)
)
Here &#39;N&#39; denotes ambigious nucleotide.
=#</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/MurrellGroup/SeededAlignment.jl/blob/429af62de75af9f9599992e8f4babe1c63d70471/src/seq_alignment/needleman_wunsch.jl#L65-L112">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SeededAlignment.read_fasta-Tuple{String}" href="#SeededAlignment.read_fasta-Tuple{String}"><code>SeededAlignment.read_fasta</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">read_fasta(filepath::String)</code></pre><p>Reads in a fasta file and returns a tuple of (seqnames, seqs).</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/MurrellGroup/SeededAlignment.jl/blob/429af62de75af9f9599992e8f4babe1c63d70471/src/utils/fasta_io.jl#L1-L5">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SeededAlignment.seed_chain_align-Tuple{BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}, BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}}" href="#SeededAlignment.seed_chain_align-Tuple{BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}, BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}}"><code>SeededAlignment.seed_chain_align</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">seed_chain_align(A::LongDNA{4}, B::LongDNA{4}; moveset::Moveset=STD_NOISY_MOVESET, scoring::ScoringScheme=STD_SCORING)</code></pre><p>(SeededAlignment wrapper - DE-NOVO)</p><p>Computes a heuristically guided global pairwise alignment of two ungapped DNA sequence <code>A</code> and <code>B</code> based on seeding heuristic. The seeds are then joined together  by computing an optimal partial alignment between seeds with the Needleman-Wunsch algorithm (nw_align). Optimal in this context meaning with repect to the choosen <code>Moveset</code> and <code>ScoringScheme</code>.</p><p>The advantage of this method is that it is much faster than nw_align and produces similar results for most usecases. </p><p><strong>Extended Help</strong></p><p><strong>Arguments</strong></p><ul><li><code>A::LongDNA{4}</code>: 1st DNA sequence to be aligned</li><li><code>B::LongDNA{4}</code>: 2nd DNA sequence to be aligned</li><li><code>moveset::Moveset=STD_NOISY_MOVESET</code>: Defines allowable alignment moves (e.g. insertions/deletions and their penalty)</li><li><code>scoring::ScoreScheme=STD_SCORING</code>: Defines alignment scoring together with moveset</li></ul><p><strong>Returns</strong></p><ul><li><code>Tuple{LongDNA{4},LongDNA{4}}</code>: Tuple representation of pairwise alignment of DNA sequences <code>A</code> and <code>B</code>.</li></ul><p><strong>Example</strong></p><pre><code class="language-julia hljs"># input sequences with no reading frame assumed
A = LongDNA{4}(&quot;AATGCTC&quot;)
B = LongDNA{4}(&quot;ACATGTC&quot;)
# produce alignment
alignment = seed_chain_align(A, B)
println(alignment)
#= resulting alignment
alignment = (
	LongDNA{4}(&quot;A-ATGCTC&quot;), 
	LongDNA{4}(&quot;ACATG-TC&quot;)
)
=#</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/MurrellGroup/SeededAlignment.jl/blob/429af62de75af9f9599992e8f4babe1c63d70471/src/seq_alignment/seed_chain_align.jl#L1-L39">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SeededAlignment.seed_chain_align-Tuple{}" href="#SeededAlignment.seed_chain_align-Tuple{}"><code>SeededAlignment.seed_chain_align</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">seed_chain_align(; 
    	ref::LongDNA{4}, 
    	query::LongDNA{4}, 
    	moveset::Moveset = STD_CODON_MOVESET, 
    	scoring::ScoringScheme = STD_SCORING,
    	codon_scoring_on::Bool = false,
    	do_clean_frameshifts::Bool = true, 
    	verbose::Bool = false)</code></pre><p>(SeededAlignment wrapper - CODING given trusted CDS anchor/reference)</p><p>Computes a heuristically guided global pairwise alignment of two ungapped CDS (Coding DNA Sequences) <code>ref</code> and <code>query</code> by using <code>ref</code> as an anchor to determine the apprioate reading frame coordinates for <code>query</code> and using a seeding heuristic for speedup. The seeds are then joined together by computing an optimal partial alignment between  seeds with the Needleman-Wunsch algorithm (nw_align). Optimal in this context means optimal with repect to the choosen <code>Moveset</code> and <code>ScoringScheme</code>.</p><p>The advantage of this method is that it is much faster than nw_align and produces similar results for most usecases. </p><p><strong>Extended Help</strong></p><p><strong>Arguments</strong></p><ul><li><code>ref::LongDNA{4}</code>: Anchored trusted CDS which decides the reading frame coordinates in the alignment</li><li><code>query::LongDNA{4}</code>: CDS (with possible frameshifts due to e.g. sequencing errors) which is aligned to <code>ref</code> and adopts its reading frame coordinates. </li><li><code>moveset::Moveset = STD_CODON_MOVESET</code>: Defines allowable alignment moves (e.g. insertions/deletions and their penalty)</li><li><code>scoring::ScoringScheme = STD_SCORING</code>: Defines alignment scoring together with moveset</li><li><code>codon_scoring_on::Bool = false</code>: Whether to apply additional scoring on codon-level </li><li><code>do_clean_frameshifts::Bool = true</code>: Whether to clean the alignment output of gaps which cause frameshifts - this produces a protein alignment on a nucleotide level. </li><li><code>verbose::Bool = false</code>: Whether to verbosely display what edits were made during the cleaning of frameshifts. </li></ul><p><strong>Returns</strong></p><ul><li><code>Tuple{LongDNA{4},LongDNA{4}}</code>: Tuple representation of pairwise alignment of DNA sequences <code>ref</code> and <code>query</code>. </li></ul><p>Note that this represents a protein alignment on a nucleotide level if <code>(do_clean_frameshifts == true)</code>.</p><p><strong>Example</strong></p><pre><code class="language-julia hljs">anchor_CDS =    LongDNA{4}(&quot;ATGCCAGTA&quot;)
# untrusted_CDS may contain some frameshift errors due to e.g. sequencing or annotation errors.
untrusted_CDS = LongDNA{4}(&quot;ATGTA&quot;) 
# frameshift errors are removed from the cleaned alignment
cleaned_CDS_alignment = seed_chain_align(ref=anchor_CDS, query=untrusted_CDS, clean_frameshifts=true)
println(cleaned_CDS_alignment)
#= resulting alignment:

cleaned_CDS_alignment = (
	LongDNA{4}(&quot;ATGCCAGTA&quot;), 
	LongDNA{4}(&quot;ATG---NTA&quot;)
)
Here &#39;N&#39; denotes ambigious nucleotide.
=#</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/MurrellGroup/SeededAlignment.jl/blob/429af62de75af9f9599992e8f4babe1c63d70471/src/seq_alignment/seed_chain_align.jl#L64-L116">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SeededAlignment.write_fasta-Tuple{String, Vector{BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}}}" href="#SeededAlignment.write_fasta-Tuple{String, Vector{BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}}}"><code>SeededAlignment.write_fasta</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">write_fasta(filepath::String, sequences::Vector{LongDNA{4}}; seq_names = nothing)</code></pre><p>Writes a fasta file from a vector of sequences, with optional seq_names.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/MurrellGroup/SeededAlignment.jl/blob/429af62de75af9f9599992e8f4babe1c63d70471/src/utils/fasta_io.jl#L14-L18">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SeededAlignment.write_fasta-Union{Tuple{N}, Tuple{String, NTuple{N, BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}}}} where N" href="#SeededAlignment.write_fasta-Union{Tuple{N}, Tuple{String, NTuple{N, BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}}}} where N"><code>SeededAlignment.write_fasta</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">write_fasta(filepath::String, sequences::Tuple{LongDNA{4},LongDNA{4}}; seq_names = nothing)</code></pre><p>Writes a fasta file from a Tuple of sequences, with optional seq_names.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/MurrellGroup/SeededAlignment.jl/blob/429af62de75af9f9599992e8f4babe1c63d70471/src/utils/fasta_io.jl#L31-L35">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SeededAlignment.STD_CODON_MOVESET" href="#SeededAlignment.STD_CODON_MOVESET"><code>SeededAlignment.STD_CODON_MOVESET</code></a> — <span class="docstring-category">Constant</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">STD_CODON_MOVESET</code></pre><p>Constants that represents the default codon moveset with frameshift moves allowed</p><p><strong>default parameter values</strong></p><pre><code class="language-julia hljs"># the tuple becomes both insertions and deletions in the moveset
const STD_CODON_MOVESET = Moveset(
    (
        Move(ref=false, step_length=1, score=-2.0, extendable=true),
        Move(ref=true,  step_length=3, score=-1.0, extendable=true)
    )
)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/MurrellGroup/SeededAlignment.jl/blob/429af62de75af9f9599992e8f4babe1c63d70471/src/alignmentOptions/moveset.jl#L146-L162">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SeededAlignment.STD_NOISY_MOVESET" href="#SeededAlignment.STD_NOISY_MOVESET"><code>SeededAlignment.STD_NOISY_MOVESET</code></a> — <span class="docstring-category">Constant</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">STD_NOISY_MOVESET</code></pre><p>Constants that represents codon oblivious moveset that favors gaps of 3. </p><p><strong>default parameter values</strong></p><pre><code class="language-julia hljs"># the tuple becomes both insertions and deletions in the moveset
const STD_NOISY_MOVESET = Moveset(
    (
        Move(ref=false, step_length=1, score=-2.0, extendable=true),
        Move(ref=true,  step_length=3, score=-1.0, extendable=true)
    )
)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/MurrellGroup/SeededAlignment.jl/blob/429af62de75af9f9599992e8f4babe1c63d70471/src/alignmentOptions/moveset.jl#L170-L186">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SeededAlignment.STD_SCORING" href="#SeededAlignment.STD_SCORING"><code>SeededAlignment.STD_SCORING</code></a> — <span class="docstring-category">Constant</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">STD_SCORING</code></pre><p>constant <code>ScoreScheme</code> that stores the default scoring parameters used in alignment methods.</p><p><strong>default parameter values</strong></p><pre><code class="language-julia hljs">const STD_SCORING = ScoringScheme(
	extension_score=-0.3,
	kmer_length=12,
	edge_ext_begin=true,
	edge_ext_end=true,
	nucleotide_mismatch_score = -0.8,
	nucleotide_match_score = 0.0,
	codon_match_bonus_score = 6.0
)
</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/MurrellGroup/SeededAlignment.jl/blob/429af62de75af9f9599992e8f4babe1c63d70471/src/alignmentOptions/scoring_scheme.jl#L144-L163">source</a></section></article></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../">« Home</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.14.1 on <span class="colophon-date" title="Saturday 23 August 2025 17:13">Saturday 23 August 2025</span>. Using Julia version 1.11.6.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
